src/block/block_addr.c

```diff
@@ -207,34 +214,38 @@ __block_ckpt_unpack(WT_SESSION_IMPL *session, WT_BLOCK *block, const uint8_t *ck
      * Passing an address cookie size of 0 so the unpack function doesn't read an object ID.
      */
     ++ckpt;
-    WT_RET(__block_addr_unpack(
-      block, &ckpt, 0, &ci->root_objectid, &ci->root_offset, &ci->root_size, &ci->root_checksum));
-    WT_RET(__block_addr_unpack(block, &ckpt, 0, &ci->alloc.objectid, &ci->alloc.offset,
+    WT_RET(__block_addr_unpack(session, block, &ckpt, 0, &ci->root_objectid, &ci->root_offset,
+      &ci->root_size, &ci->root_checksum));
+    WT_RET(__block_addr_unpack(session, block, &ckpt, 0, &ci->alloc.objectid, &ci->alloc.offset,
       &ci->alloc.size, &ci->alloc.checksum));
-    WT_RET(__block_addr_unpack(block, &ckpt, 0, &ci->avail.objectid, &ci->avail.offset,
+    WT_RET(__block_addr_unpack(session, block, &ckpt, 0, &ci->avail.objectid, &ci->avail.offset,
       &ci->avail.size, &ci->avail.checksum));
-    WT_RET(__block_addr_unpack(block, &ckpt, 0, &ci->discard.objectid, &ci->discard.offset,
+    WT_RET(__block_addr_unpack(session, block, &ckpt, 0, &ci->discard.objectid, &ci->discard.offset,
       &ci->discard.size, &ci->discard.checksum));
     WT_RET(__wt_vunpack_uint(&ckpt, 0, &a));
     ci->file_size = (wt_off_t)a;
     WT_RET(__wt_vunpack_uint(&ckpt, 0, &a));
     ci->ckpt_size = a;
-    ci->root_objectid = ci->alloc.objectid = ci->avail.objectid = ci->discard.objectid = 0;
-    if (WT_PTRDIFF(ckpt, start) != ckpt_size) {
-        WT_RET(__wt_vunpack_uint(&ckpt, 0, &a));
-        ci->root_objectid = ci->alloc.objectid = ci->avail.objectid = ci->discard.objectid =
-          (uint32_t)a;
+
+    /* The first part of the checkpoint cookie is optionally followed by an object ID. */
+    objectid = 0;
+    flags = 0;
+    if (WT_PTRDIFF(ckpt, begin) != ckpt_size) {
+        flags = *ckpt++;
+        if (LF_ISSET(WT_BLOCK_COOKIE_FILEID)) {
+            WT_RET(__wt_vunpack_uint(&ckpt, 0, &a));
+            objectid = (uint32_t)a;
+        }
     }
+    ci->root_objectid = ci->alloc.objectid = ci->avail.objectid = ci->discard.objectid = objectid;
 
     /*
-     * Assert the entire cookie was consumed.
-     *
-     * TODO: The assert prevents future extensions of the cookie this code should ignore, but that
-     * may be unavoidable, as the object ID is not a required field, extending the cookie may not be
-     * simple. Consider adding a flag byte before the optional object ID, describing the object ID
-     * and any other fields.
+     * If there is no flag value or the flag value is the object ID flag by itself, assert the
+     * cookie was entirely consumed. Future extensions will use different cookie flag values
+     * (although the file-ID flag might still be set, our test is for equality).
      */
-    WT_ASSERT(session, WT_PTRDIFF(ckpt, start) == ckpt_size);
+    WT_ASSERT(session,
+      (flags != 0 && flags != WT_BLOCK_COOKIE_FILEID) || WT_PTRDIFF(ckpt, begin) == ckpt_size);
```

Comment:
This doesn't seem right to me.  Isn't `WT_PTRDIFF(ckpt, start) == ckpt_size)` always true?

Reply:
If the flags value isn't one we know about, then it's a future release and we may not have consumed the fields added by that release.

Let me know if I'm missing something.